### DISCUSSION TITLE
Pratt parsing with extra metadata

### DISCUSSION THREAD

**Discussion by catapillie** [zesterer/chumsky#617] [Q&A] [answered]
Hello!
I've been using Pratt parsing for unary and binary operations, and it's been really useful. However I haven't found a way to propagate extra metadata when folding (similarly to how `map_with`, `foldr_with` and `foldl_with` work). I'd like to be able to fold the operations while retrieving the span of the currently parsed operation.
```
  a    *    1    +    -x
  ^         ^          ^     span of each atom
                      ^^     span of the negation
  ^^^^^^^^^^^                span of the product
  ^^^^^^^^^^^^^^^^^^^^^^     span of the sum
```

There doesn't seem to be a way to directly do this using `pratt::prefix`, `pratt::postfix`, etc... Is this currently not possible?
On `1.0.0-alpha.6`. Thanks in advance!

**Comment by zesterer** [comment, marked as answer]
It's easier than you might think! You can simply add the span to the output type of the operator / atom like so:

```rs
atom.map_with(|x, e| (x, e.span())).pratt((
    infix(left(1), my_op.map_with(|x, e| (x, e.span())), |l, op, r, e| {
        // Note that in here, `l.1`, `op.1`, and `r.1` all give the span of their respective elements
        (Expr::MyOp(l, op, r), e.span())
    }),
))
```

**Reply by catapillie** [reply]
Thanks a lot!
